scripts/trace-dd.mjs

```diff
@@ -0,0 +1,178 @@
+  return filename
+}
+
+const getPackageName = (filename) => {
+  const match = /.+[\\/]node_modules[\\/]((?:@[^\\/]+[\\/])?[^\\/]+)/.exec(
+    cleanFilename(filename)
+  )
+  return match && match[1]
+}
+
+/**
+ * Create, reports spans recursively with its inner child spans.
+ */
+const reportSpanRecursively = (tracer, trace, parentSpan) => {
+  // build-* span contains tags with path to the modules, trying to clean up if possible
+  const isBuildModule = trace.name.startsWith('build-module-')
+  if (isBuildModule) {
+    trace.packageName = getPackageName(trace.tags.name)
+    // replace name to cleaned up pkg name
+    trace.tags.name = trace.packageName
+    if (trace.children) {
+      const queue = [...trace.children]
+      trace.children = []
+      for (const e of queue) {
+        if (e.name.startsWith('build-module-')) {
+          const pkgName = getPackageName(e.tags.name)
+          if (!trace.packageName || pkgName !== trace.packageName) {
+            trace.children.push(e)
+          } else {
+            if (e.children) queue.push(...e.children)
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * interface TraceEvent {
+   *  traceId: string;
+   *  parentId: number;
+   *  name: string;
+   *  id: number;
+   *  startTime: number;
+   *  timestamp: number;
+   *  duration: number;
+   *  tags: Record<string, any>
+   * }
+   */
+  let span = tracer.startSpan(trace.name, {
+    startTime: trace.startTime,
+    childOf: parentSpan,
+    tags: Object.keys(trace?.tags).length > 0 ? trace?.tags : undefined,
+  })
+
+  // Spans should be reported in chronological order
+  trace.children?.sort((a, b) => a.startTime - b.startTime)
+  trace.children?.forEach((childTrace) =>
+    reportSpanRecursively(tracer, childTrace, span)
+  )
+
+  span.finish(trace.startTime + trace.duration / 1000)
+  return span
+}
+
+/**
+ * Read generated trace from file system, augment & sent it to the remote tracer.
+ */
+const collectTraces = async (filePath, metadata) => {
+  const tracer = Tracer.init({
+    tags: metadata,
+    // Setting external env variable `DD_TRACE_DEBUG=true` will emit this log
+    logLevel: 'error',
+    // TODO: this is due to generated trace have excessive numbers of spans
+    // for build-module-*, using default flush causes overflow to the agent.
+    flushInterval: 20,
+    flushMinSpans: 10,
+  })
+
+  const readLineInterface = createInterface({
+    input: createReadStream(filePath),
+    crlfDelay: Infinity,
+  })
+
+  const traces = new Map()
+  const rootTraces = []
+
+  // Input trace file contains newline-seperated sets of traces, where each line is valid JSON
+  // type of Array<TraceEvent>. Read it line-by-line to manually reconstruct trace trees.
+  //
+  // We have to read through end of the trace -
+  // There's no gaurantee collected trace have order for its parent span to inner childs, which
+  // we have to reconstruct all of tree here before send actual traces.
```

Comment:
The wording of this sentence is a little confusing. Maybe something like: "Trace events in the input file can appear out of order, so we need to remodel the shape of the span tree before reporting."